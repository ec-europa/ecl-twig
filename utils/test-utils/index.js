const Twig = require('twig');
const { format } = require('prettier');

const renderTwigFile = (file, options, cb) =>
  Twig.renderFile(file, options, (err, html) =>
    cb(err, format(html, { parser: 'html' }))
  );

const renderTwigFileAsNode = (file, options) =>
  new Promise((resolve, reject) =>
    Twig.renderFile(file, options, (err, html) => {
      if (err) return reject(err);

      const div = document.createElement('div');
      div.innerHTML = html.trim();

      return resolve(div.firstChild);
    })
  );

module.exports = {
  renderTwigFile,
  renderTwigFileAsNode,
};
