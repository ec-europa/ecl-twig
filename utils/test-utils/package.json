{
  "name": "@ecl-twig/test-utils",
  "version": "2.29.0",
  "description": "A local test helper",
  "main": "index.js",
  "private": true,
  "devDependencies": {
    "deepmerge": "4.2.2"
  }
}
