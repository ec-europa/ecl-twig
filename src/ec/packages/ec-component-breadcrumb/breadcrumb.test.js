import path from 'path';
import { merge, renderTwigFileAsNode } from '@ecl-twig/test-utils';

import breadcrumbDataSimple from '@ecl/ec-specs-breadcrumb/demo/data-simple';
import breadcrumbDataLong from '@ecl/ec-specs-breadcrumb/demo/data';

function formatLink(l) {
  const link = {
    label: l.label,
    path: l.href,
  };

  return link;
}

describe('EC - Breadcrumb', () => {
  const template = path.resolve(__dirname, './breadcrumb.html.twig');
  const render = params => renderTwigFileAsNode(template, params);
  const defaultIconPath = 'static/icons.svg';

  describe('Simple', () => {
    const data = {
      links: breadcrumbDataSimple.items.map(formatLink),
      icon_file_path: defaultIconPath,
      navigation_text: breadcrumbDataSimple.label,
      ellipsis_label: 'Click to expand',
    };

    test(`renders correctly`, () => {
      expect.assertions(1);
      return expect(render(data)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const optionsWithExtraClasses = merge(data, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(optionsWithExtraClasses)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra attributes', () => {
      expect.assertions(1);

      const optionsWithExtraClasses = merge(data, {
        extra_attributes: [
          { name: 'data-test', value: 'data-test-value' },
          { name: 'data-test-1', value: 'data-test-value-1' },
        ],
      });

      return expect(render(optionsWithExtraClasses)).resolves.toMatchSnapshot();
    });
  });

  describe('Long', () => {
    const data = {
      links: breadcrumbDataLong.items.map(formatLink),
      icon_file_path: defaultIconPath,
      navigation_text: breadcrumbDataLong.label,
      ellipsis_label: 'Click to expand',
    };

    test(`renders correctly`, () => {
      expect.assertions(1);
      return expect(render(data)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const optionsWithExtraClasses = merge(data, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(optionsWithExtraClasses)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra attributes', () => {
      expect.assertions(1);

      const optionsWithExtraClasses = merge(data, {
        extra_attributes: [
          { name: 'data-test', value: 'data-test-value' },
          { name: 'data-test-1', value: 'data-test-value-1' },
        ],
      });

      return expect(render(optionsWithExtraClasses)).resolves.toMatchSnapshot();
    });
  });
});
