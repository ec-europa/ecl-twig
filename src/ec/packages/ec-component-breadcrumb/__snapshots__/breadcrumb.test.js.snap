// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EC - Site Header Breadcrumb long - renders correctly 1`] = `
<nav
  aria-label="You are here:"
  class="ecl-breadcrumb"
  data-ecl-breadcrumb="true"
>
  <ol
    class="ecl-breadcrumb__container"
  >
    <li
      aria-hidden="false"
      class="ecl-breadcrumb__segment"
      data-ecl-breadcrumb-item="static"
    >
      <a
        class="ecl-link ecl-link--standalone ecl-breadcrumb__link"
        href="/example"
      >
        Home
      </a>
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-icon--inverted ecl-breadcrumb__icon"
        focusable="false"
        role="presentation"
      >
        <use
          xlink:href="static/icons.svg#ui--corner-arrow"
        />
      </svg>
    </li>
    <li
      aria-hidden="false"
      class="ecl-breadcrumb__segment ecl-breadcrumb__segment--ellipsis"
      data-ecl-breadcrumb-ellipsis="true"
    >
      <button
        aria-label="Click to expand"
        class="ecl-breadcrumb__ellipsis"
        data-ecl-breadcrumb-ellipsis-button="true"
        type="button"
      >
        â€¦
      </button>
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-icon--inverted ecl-breadcrumb__icon"
        focusable="false"
        role="presentation"
      >
        <use
          xlink:href="static/icons.svg#ui--corner-arrow"
        />
      </svg>
    </li>
    <li
      aria-hidden="true"
      class="ecl-breadcrumb__segment"
      data-ecl-breadcrumb-item="expandable"
    >
      <a
        class="ecl-link ecl-link--standalone ecl-breadcrumb__link"
        href="/example"
      >
        About the European Commission
      </a>
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-icon--inverted ecl-breadcrumb__icon"
        focusable="false"
        role="presentation"
      >
        <use
          xlink:href="static/icons.svg#ui--corner-arrow"
        />
      </svg>
    </li>
    <li
      aria-hidden="true"
      class="ecl-breadcrumb__segment"
      data-ecl-breadcrumb-item="expandable"
    >
      <a
        class="ecl-link ecl-link--standalone ecl-breadcrumb__link"
        href="/example"
      >
        Organisational structure
      </a>
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-icon--inverted ecl-breadcrumb__icon"
        focusable="false"
        role="presentation"
      >
        <use
          xlink:href="static/icons.svg#ui--corner-arrow"
        />
      </svg>
    </li>
    <li
      aria-hidden="false"
      class="ecl-breadcrumb__segment"
      data-ecl-breadcrumb-item="static"
    >
      <a
        class="ecl-link ecl-link--standalone ecl-breadcrumb__link"
        href="/example"
      >
        How the Commission is organised
      </a>
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-icon--inverted ecl-breadcrumb__icon"
        focusable="false"
        role="presentation"
      >
        <use
          xlink:href="static/icons.svg#ui--corner-arrow"
        />
      </svg>
    </li>
    <li
      aria-current="page"
      aria-hidden="false"
      class="ecl-breadcrumb__segment ecl-breadcrumb__current-page"
      data-ecl-breadcrumb-item="static"
    >
      News
    </li>
  </ol>
</nav>
`;

exports[`EC - Site Header Breadcrumb simple - renders correctly 1`] = `
<nav
  aria-label="You are here:"
  class="ecl-breadcrumb"
  data-ecl-breadcrumb="true"
>
  <ol
    class="ecl-breadcrumb__container"
  >
    <li
      aria-hidden="false"
      class="ecl-breadcrumb__segment"
      data-ecl-breadcrumb-item="static"
    >
      <a
        class="ecl-link ecl-link--standalone ecl-breadcrumb__link"
        href="/example"
      >
        Home
      </a>
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-icon--inverted ecl-breadcrumb__icon"
        focusable="false"
        role="presentation"
      >
        <use
          xlink:href="static/icons.svg#ui--corner-arrow"
        />
      </svg>
    </li>
    <li
      aria-hidden="false"
      class="ecl-breadcrumb__segment"
      data-ecl-breadcrumb-item="static"
    >
      <a
        class="ecl-link ecl-link--standalone ecl-breadcrumb__link"
        href="/example"
      >
        About the European Commission
      </a>
      <svg
        aria-hidden="true"
        class="ecl-icon ecl-icon--xs ecl-icon--rotate-90 ecl-icon--inverted ecl-breadcrumb__icon"
        focusable="false"
        role="presentation"
      >
        <use
          xlink:href="static/icons.svg#ui--corner-arrow"
        />
      </svg>
    </li>
    <li
      aria-current="page"
      aria-hidden="false"
      class="ecl-breadcrumb__segment ecl-breadcrumb__current-page"
      data-ecl-breadcrumb-item="static"
    >
      News
    </li>
  </ol>
</nav>
`;
